// mincloud/lib/commands/new.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.cli = undefined;

var _utils = require('../utils');

var _mkdirp = require('mkdirp');

var _mkdirp2 = _interopRequireDefault(_mkdirp);

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _fs = require('fs');

var _fs2 = _interopRequireDefault(_fs);

var _util = require('util');

var _util2 = _interopRequireDefault(_util);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _asyncToGenerator(fn) { return function () { var gen = fn.apply(this, arguments); return new Promise(function (resolve, reject) { function step(key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { return Promise.resolve(value).then(function (value) { step("next", value); }, function (err) { step("throw", err); }); } } return step("next"); }); }; }

var mkdir = _util2.default.promisify(_mkdirp2.default);
var echo = _util2.default.promisify(_fs2.default.writeFile);

var FUNCTION_TEMPLATE = 'exports.main = function functionName(event, callback) {\n  callback(null, "hello world")\n}';

var cli = exports.cli = (0, _utils.ensureAuth)(function () {
  var _ref = _asyncToGenerator( /*#__PURE__*/regeneratorRuntime.mark(function _callee(engine, functionName) {
    var rootFolder = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : './';
    var remark, target, targetfile;
    return regeneratorRuntime.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            if (functionName) {
              _context.next = 2;
              break;
            }

            throw (0, _utils.usageError)('缺少必填字段 <function_name>', '', '用法：', '    ' + engine.config.get('prefix') + ' new <function_name> [cloud_function_root] [-m remark]');

          case 2:
            remark = engine.config.get('message');
            _context.next = 5;
            return engine.request({
              uri: '/oserve/v1.3/cloud-function/',
              method: 'POST',
              json: {
                name: functionName,
                function_code: FUNCTION_TEMPLATE,
                remark: remark
              }
            });

          case 5:
            target = _path2.default.resolve(rootFolder, functionName);
            targetfile = _path2.default.join(target, 'index.js');
            _context.next = 9;
            return mkdir(target);

          case 9:
            _context.next = 11;
            return echo(targetfile, FUNCTION_TEMPLATE);

          case 11:

            console.log('创建成功');
            console.log('');
            console.log(target);
            console.log(targetfile);
            console.log('');
            console.log('- \u51FD\u6570\u540D\uFF1A' + functionName);
            console.log('- \u51FD\u6570\u6839\u76EE\u5F55: ' + rootFolder);

          case 18:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, undefined);
  }));

  return function (_x, _x2) {
    return _ref.apply(this, arguments);
  };
}());
